File: 2024/day_02.py
import argparse

from pathlib import Path
from time import time

def _parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--part", "-p",
        type=int,
        choices={1, 2},
        help="Set puzzle part"
    )
    args = parser.parse_args()
    if not args.part:
        parser.error("Which part are you solving?")
    return args

def is_safe(report: list) -> bool:
    prev_d = None
    for i in range(1, len(report)):
        d = report[i] - report[i-1]
        if not d or abs(d) > 3:
            return False
        if prev_d and prev_d * d < 0:
            return False
        prev_d = d
    return True

def is_really_safe(report: list) -> bool:
    if is_safe(report):
        return True
    for i in range(len(report)):
        if is_safe(report[:i] + report[i + 1:]):
            return True
    return False

if __name__ == "__main__":
    args = _parse_args()
    t = time()
    with Path(f"{Path(__file__).parent}/inputs/{Path(__file__).stem}.txt").open("r") as file:
        data = [[int(x) for x in line.split(" ")] for line in file.read().split("\n")]
    if args.part == 1:
        print(sum(is_safe(report) for report in data))
    else:
        print(sum(is_really_safe(report) for report in data))
    print(time() - t)

File: 2021/day_02.rb
require 'optparse'
require 'ostruct'

def parse_args
  args = OpenStruct.new
  OptionParser.new do |arg|
    arg.on('-p', '--part PART', 'The part you\'re solving') { |a| args.part = a }
  end.parse!
  if not args.part
    raise "Which part are you solving?"
  end
  return args
end

def parse_input(line)
  movement = OpenStruct.new
  value = line.scan(/\d+/)[0].to_i
  if line.include? "forward"
    movement.x = value
    movement.y = 0
  else
    movement.x = 0
    movement.y = line.include?("down") ? value : -value
  end
  return movement
end

if __FILE__ == $0
    args = parse_args
    t = Time.now
    data = nil
    File.open("#{File.dirname($0)}/inputs/#{File.basename($0, ".*")}.txt", "r") do |f|
      data = f.read.split("\n").map{ |line| parse_input(line)}
    end
    pos = OpenStruct.new({ x: 0, y: 0 })
    if args.part == "1"
      data.each {
        |mvt|
        pos.x += mvt.x
        pos.y += mvt.y
      }
    else
      aim = 0
      data.each {
        |mvt|
        pos.x += mvt.x
        pos.y += mvt.x * aim
        aim += mvt.y
      }
    end
    puts pos.x * pos.y
    puts Time.now - t
end
